segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java

```
@@ -106,19 +163,131 @@ public void stop() {
+            try {
+                SSLContext context = SSLContext.getInstance("TLS");
+                TrustManagerFactory trustManager = getTrustManager(trustStore, trustStorePasswordPath);
+                KeyManagerFactory keyFactory = getKeyManager(keyStore, keyStorePasswdPath);
+
+                context.init(keyFactory.getKeyManagers(), trustManager.getTrustManagers(), null);
+                SSLSocketFactory factory = context.getSocketFactory();
+
+                Socket sock = context.getSocketFactory().createSocket(new Socket(host, port), host, port, true);
+                BufferedReader reader = null;
+                try {
+                    OutputStream outstream = sock.getOutputStream();
+                    outstream.write("stat".getBytes());
+                    outstream.flush();
+
+                    reader =
+                            new BufferedReader(
+                                    new InputStreamReader(sock.getInputStream()));
+                    String line = reader.readLine();
+                    if (line != null && line.startsWith("Zookeeper version:")) {
+                        log.info("Server UP");
+                        return true;
+                    }
+                } catch (Exception e) {
+                  e.printStackTrace();
+                } finally {
+                    sock.close();
+                    if (reader != null) {
+                        reader.close();
+                    }
+                }
+            } catch (IOException | CertificateException | NoSuchAlgorithmException | KeyStoreException
+                    | KeyManagementException | UnrecoverableKeyException e) {
+                // ignore as this is expected
+                log.info("server " + address + " not up " + e);
+            }
+
+            if (MathUtils.now() > start + timeout) {
+                break;
+            }
+            try {
+                Thread.sleep(250);
+            } catch (InterruptedException e) {
```

I'm wondering about catching `InterruptedException` here, should we let it bubble up?

---

Fixed. Bubbling this up as a `RunTimeException`.

---

Use `Exceptions.handleInterrupted`